src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java

```diff
@@ -622,71 +626,138 @@ private AIState waitForShovel()
+    }*/
 
     /**
-     * Ensures that we have a tool available.
+     * Ensures that we have a appropriate tool available.
+     * Will set {@code needsTool} accordingly.
      *
-     * @param tool tool required for block
-     * @return true if we have a tool
+     * @param toolType type of tool we check for.
+     * @return true if we have the tool
      */
-    private boolean checkForTool(@NotNull final String tool)
+    protected boolean checkForTool(@NotNull final ToolType toolType)
     {
-        final boolean needsTool = !InventoryFunctions
-                                           .matchFirstInProvider(
-                                                   worker,
-                                       stack -> Utils.isTool(stack, tool));
+        return checkForTool(toolType, TOOL_LEVEL_WOOD_OR_GOLD);
+    }
 
-        final int hutLevel = worker.getWorkBuilding().getBuildingLevel();
-        final InventoryCitizen inventory = worker.getInventoryCitizen();
-        final boolean isUsable = InventoryUtils.isToolInItemHandler(new InvWrapper(inventory), tool, hutLevel);
+    protected boolean checkForTool(@NotNull final ToolType toolType, final int minimalLevel)
+    {
+        if (checkForNeededTool(toolType, minimalLevel))
+        {
+            getOwnBuilding().setNeedsTool(toolType, minimalLevel);
+        }
+        else if(getOwnBuilding().needsTool(toolType))
+        {
+            getOwnBuilding().setNeedsTool(ToolType.NONE, TOOL_LEVEL_HAND);
+        }
 
+        return getOwnBuilding().needsTool(toolType);
+    }
 
-        if (!needsTool && isUsable)
+    /**
+     * Check if we need a tool.
+     *
+     * Do not use it to find a pickaxe as it need a minimum level
+     * @param tool tool required for block
+     * @return true if we need a tool
+     */
+    private boolean checkForNeededTool(@NotNull final ToolType toolType, final int minimalLevel)
+    {
+        final int maxToolLevel = worker.getWorkBuilding().getMaxToolLevel();
+        final InventoryCitizen inventory = worker.getInventoryCitizen();
+        if (InventoryUtils.isToolInItemHandler(new InvWrapper(inventory), toolType, minimalLevel, maxToolLevel))
         {
             return false;
         }
+
         delay += DELAY_RECHECK;
         if (walkToBuilding())
         {
             return true;
         }
-        if (isToolInHut(tool))
+        if (isToolInHut(toolType, minimalLevel))
         {
             return false;
         }
         if (!getOwnBuilding().hasOnGoingDelivery())
         {
-            chatSpamFilter.talkWithoutSpam(COM_MINECOLONIES_COREMOD_ENTITY_WORKER_TOOLREQUEST, tool, InventoryUtils.swapToolGrade(hutLevel));
+            chatRequestTool(toolType, minimalLevel, maxToolLevel);
+
         }
         return true;
     }
 
+    private void chatRequestTool(@NotNull final ToolType toolType, final int minimalLevel, final int maximumLevel)
+    {
+        switch(toolType)
+        {
+            case PICKAXE:
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Reduce this switch case number of lines from 17 to at most 10, for example by extracting code into methods. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1151)